ddtrace/bootstrap/sitecustomize.py

```diff
@@ -111,7 +188,20 @@ def update_patched_modules():
         update_patched_modules()
         from ddtrace import patch_all
 
+        # We need to clean up after we have imported everything we need from
+        # ddtrace, but before we register the patch-on-import hooks for the
+        # integrations. This is because registering a hook for a module
+        # that is already imported causes the module to be patched immediately.
+        # So if we unload the module after registering hooks, we effectively
+        # remove the patching, thus breaking the tracer integration.
+        cleanup_loaded_modules()
```

Comment:
is there a finite set of modules which are imported at this point?

are there any chances the import list is not well defined up front? e.g. could any integration code be imported, which imports an array of modules?


is there a way for us to test (in CI) that only the expected modules exist in `sys.modules`?

Reply:
@brettlangdon I'm not sure I understand the need for this approach given the generic approach this change takes to module cloning. Can you help me understand an example of a case where `sys.modules` might be in an unexpected state?

> is there a way for us to test (in CI) that only the expected modules exist in sys.modules?

Certainly, given that we have a clear definition of "expected" in this context.

Reply:
this could just be partial ignorance on my part.

do we know exactly which modules will be loaded by this point?

can that list of loaded modules ever change?

since this is happening in sitecustomize, I'd expect this is all running before the user's code, so therefore the list of loaded modules should be pretty static based on our dependencies.

if so, and the list is pretty much set in stone ("ish"), then that limits the overall risk factor because we probably won't need to chase down and continually add to the deny list.

so my question about testing was mostly... is it valuable to try and codify this expectation in our tests? if someone were to add a new dependency, would we know if this breaks? or if a specific version of a dependency changes (or a deps dep) could that have impact on how this works?